cmake/common/check_configuration.cmake

```diff
@@ -12,155 +12,131 @@
-    elseif(MSVC OR MSVC_IDE)
-        check_cxx_compiler_flag(/std:c++20 SUPPORTS_CXX20)
-        set(HAVE_CXX20 0)
-        set(HAVE_CXX17 0)
-        set(HAVE_CXX14 0)
-        set(HAVE_CXX11 0)
-        set(HAVE_CXX0X 0)
-        if(SUPPORTS_CXX20 AND (NOT FORCE_CXX OR "${FORCE_CXX}" STREQUAL "20"))
-            set(FASTDDS_REQUIRED_FLAGS "/std:c++20 ${FASTDDS_REQUIRED_FLAGS}")
-            add_compile_options($<$<COMPILE_LANGUAGE:CXX>:/std:c++20>)
-            set(HAVE_CXX20 1)
-            set(HAVE_CXX17 1)
-            set(HAVE_CXX14 1)
-            set(HAVE_CXX11 1)
-            set(HAVE_CXX0X 1)
-        elseif(NOT SUPPORTS_CXX20 AND FORCE_CXX AND "${FORCE_CXX}" STREQUAL "20")
-            message(FATAL_ERROR "Force to support stdc++20 but not supported by the compiler")
-        else()
-            check_cxx_compiler_flag(/std:c++17 SUPPORTS_CXX17)
-            if(SUPPORTS_CXX17 AND (NOT FORCE_CXX OR "${FORCE_CXX}" STREQUAL "17"))
-                set(FASTDDS_REQUIRED_FLAGS "/std:c++17 ${FASTDDS_REQUIRED_FLAGS}")
-                add_compile_options($<$<COMPILE_LANGUAGE:CXX>:/std:c++17>)
-                set(HAVE_CXX17 1)
-                set(HAVE_CXX14 1)
-                set(HAVE_CXX11 1)
-                set(HAVE_CXX0X 1)
-            elseif(NOT SUPPORTS_CXX17 AND FORCE_CXX AND "${FORCE_CXX}" STREQUAL "17")
-                message(FATAL_ERROR "Force to support stdc++17 but not supported by the compiler")
-            else()
-                check_cxx_compiler_flag(/std:c++14 SUPPORTS_CXX14)
-                if(SUPPORTS_CXX14 AND (NOT FORCE_CXX OR "${FORCE_CXX}" STREQUAL "14"))
-                    set(FASTDDS_REQUIRED_FLAGS "/std:c++14 ${FASTDDS_REQUIRED_FLAGS}")
-                    add_compile_options($<$<COMPILE_LANGUAGE:CXX>:/std:c++14>)
-                    set(HAVE_CXX14 1)
-                    set(HAVE_CXX11 1)
-                    set(HAVE_CXX0X 1)
-                elseif(NOT SUPPORTS_CXX14 AND FORCE_CXX AND "${FORCE_CXX}" STREQUAL "14")
-                    message(FATAL_ERROR "Force to support stdc++14 but not supported by the compiler")
-                else()
-                    message(STATUS "The specified C++${FORCE_CXX} option is not supported using compiler default.")
-                    set(HAVE_CXX11 1)
-                    set(HAVE_CXX0X 1)
-                endif() # C++14
-            endif() # C++17
-        endif() # C++20
-    else()
-        set(HAVE_CXX11 0)
-        set(HAVE_CXX0X 0)
-    endif()
-endmacro()
+# get_set_stdcxx() checks and sets the compiler standard level
+#   stdversion: the level of standard fullfilment
+#   stdfeature: the feature name associated with that level
+#   gcc_flag:   the gcc/clang flag for fallback
+#   cl_flag:    the cl flag for fallback
+#   force:      if the user should enforced this standard level or not (FORCE_CXX)
+#   result:     a return variable to check if this level is available
+
+function(get_set_stdcxx stdversion stdfeature gcc_flag cl_flag force result)
```

Comment:
Welcome to *CMake world*. A world where functions cannot return a result but by setting variables in the parent scope. `get_set_stdcxx` result is used in the next `if` statement.
If the user sets the `FORCE_CXX` variable to anything other than *23, 20, 17, 14, 1Y, 11* (like an empty string) it will get a full fledge compiler.